op-chain-ops/cmd/op-version-check/main.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/mattn/go-isatty"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/exp/maps"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/upgrades"
+	"github.com/ethereum-optimism/optimism/op-service/jsonutil"
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+)
+
+type Contract struct {
+	Version string             `yaml:"version"`
+	Address superchain.Address `yaml:"address"`
+}
+
+type ChainVersionCheck struct {
+	Name      string              `yaml:"name"`
+	ChainID   uint64              `yaml:"chain_id"`
+	Contracts map[string]Contract `yaml:"contracts"`
+}
+
+func main() {
+	log.Root().SetHandler(log.StreamHandler(os.Stderr, log.TerminalFormat(isatty.IsTerminal(os.Stderr.Fd()))))
+
+	app := &cli.App{
+		Name:  "op-version-check",
+		Usage: "Determine which contract versions are deployed for chains in a superchain",
+		Flags: []cli.Flag{
+			&cli.StringSliceFlag{
+				Name:    "l1-rpc-urls",
+				Usage:   "L1 RPC URLs, the chain ID will be used to determine the superchain",
+				EnvVars: []string{"L1_RPC_URL"},
+			},
+			&cli.StringSliceFlag{
+				Name:    "l2-rpc-urls",
+				Usage:   "L2 RPC URLs, corresponding to chains to check versions for. Corresponds to all chains if empty",
+				EnvVars: []string{"L1_RPC_URL"},
+			},
+			&cli.PathFlag{
+				Name:    "outfile",
+				Usage:   "The file to write the output to. If not specified, output is written to stdout",
+				EnvVars: []string{"OUTFILE"},
+			},
+		},
+		Action: entrypoint,
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Crit("error op-version-check", "err", err)
+	}
+}
+
+// entrypoint contains the main logic of the script
+func entrypoint(ctx *cli.Context) error {
+	l1RPCURLs := ctx.StringSlice("l1-rpc-urls")
+	l2RPCURLs := ctx.StringSlice("l2-rpc-urls")
+
+	var l2ChainIDs []uint64
+
```

Comment:
For each L2 RPC:
  Call `version()` on each versioned predeploy
For each L1 RPC:
  Grab the contract addresses from superchain registry
  For each contract address:
    Call `version()`